Subaru pre-global: add support for Subaru Legacy 2015-18

martinl opened the pull request:
**Checklist**

- [x] added to README
- [x] test route added to [test_car_models](../../selfdrive/test/test_car_models.py)
- [x] route with openpilot: df5ca7660000fba8|2020-06-16--17-37-19
- [x] route with stock system: df5ca7660000fba8|2020-07-01--06-53-52
- [x] add dashcamOnly

Splitting Subaru pre-global models PR https://github.com/commaai/openpilot/pull/1489 to 4 smaller PR-s by model. This is first one, ready for review

martinl commented:
Rebased, added one more Legacy 2018 FPv1 and Ascent support from upstream

adeebshihadeh commented on selfdrive/car/subaru/values.py:
```diff
@@ -8,6 +8,7 @@ class CAR:
   ASCENT = "SUBARU ASCENT LIMITED 2019"
   IMPREZA = "SUBARU IMPREZA LIMITED 2019"
   FORESTER = "SUBARU FORESTER 2019"
+  LEGACY_2015 = "SUBARU LEGACY 2015 - 2018"
```
Can you remove the model year?

martinl replied:
My current logic is to use model years for older models, I think 2019 or 2020 Legacy on Global platform will be without model year

adeebshihadeh replied:
Let's make the convention this: global cars are just the model names and pre-global get a PREGLOBAL suffix

martinl replied:
done

adeebshihadeh commented on selfdrive/car/subaru/subarucan.py:
```diff
@@ -1,19 +1,36 @@
+      "SET_1": 1
+    }
+
+  if car_fingerprint == CAR.LEGACY_2015:
```
I'd just make a list of all the legacy cars in values.py and check if the car's in the list.

martinl replied:
done

adeebshihadeh commented on selfdrive/car/subaru/carcontroller.py:
```diff
@@ -1,32 +1,34 @@
+    self.STEER_STEP = 2                  # how often we update the steer cmd
+    self.STEER_DELTA_UP = 50             # torque increase per refresh, 0.8s to max
+    self.STEER_DELTA_DOWN = 70           # torque decrease per refresh
+    if car_fingerprint in [CAR.ASCENT, CAR.FORESTER, CAR.IMPREZA]:
+      self.STEER_DRIVER_ALLOWANCE = 60   # allowed driver torque before start limiting
+    if car_fingerprint == CAR.LEGACY_2015:
```
same thing about maintaining a list of pre-global cars

martinl replied:
done

adeebshihadeh commented on selfdrive/car/subaru/values.py:
```diff
@@ -46,4 +63,8 @@ class CAR:
+  CAR.LEGACY_2015: dbc_dict('subaru_outback_2015_generated', None),
 }
+
+GLOBAL_CAR = [CAR.ASCENT, CAR.IMPREZA, CAR.FORESTER]
+LEGACY_CAR = [CAR.LEGACY_2015]
```
Let's call it `PREGLOBAL_CARS`, and we only need one list (the global cars are just all the subarus that aren't pre-global).

martinl replied:
done

mkrupczak3 commented:
👀

adeebshihadeh commented on selfdrive/car/subaru/carstate.py:
```diff
@@ -51,18 +51,30 @@ def update(self, cp, cp_cam):
-      ret.cruiseState.speed *= CV.MPH_TO_KPH
+
+    # UDM Legacy: mph = 0
+    if self.car_fingerprint == CAR.LEGACY_PREGLOBAL:
```
Is this for all pre-global or just the legacy?

martinl replied:
This is only for Legacy and Forester, Outback has different value for mph. The ACC speed unit values are based on data from about 10 cars from different markets around the world. So far, it looks like Subaru has made the ACC speed unit more complex than average and is using different values depending on models and markets on the same platform

martinl replied:
I think another option would be removing ACC speed signal from carstate for both global and preglobal until we have a clean and reliable way to get the ACC speed unit

adeebshihadeh commented on selfdrive/car/subaru/carstate.py:
```diff
@@ -51,18 +51,30 @@ def update(self, cp, cp_cam):
 
-    self.es_distance_msg = copy.copy(cp_cam.vl["ES_Distance"])
-    self.es_lkas_msg = copy.copy(cp_cam.vl["ES_LKAS_State"])
+    if self.car_fingerprint not in PREGLOBAL_CARS:
```
All the nots make it harder to read. I think it's fine if they all start with the preglobals

martinl replied:
I'll revert back to preglobal first

martinl commented:
Reverted to using in PREGLOBAL_CARS

adeebshihadeh commented on selfdrive/car/subaru/carcontroller.py:
```diff
@@ -1,30 +1,34 @@
+    self.STEER_DELTA_UP = 50             # torque increase per refresh, 0.8s to max
+    self.STEER_DELTA_DOWN = 70           # torque decrease per refresh
+    if car_fingerprint in PREGLOBAL_CARS:
+      self.STEER_DRIVER_ALLOWANCE = 75   # allowed driver torque before start limiting
```
why is this different for pre-global cars? seems like they share the same EPS

martinl replied:
yes, looks like they can be unified to use Global's 60, since both scales are similar. I will change it

martinl replied:
Done

adeebshihadeh commented on selfdrive/car/subaru/values.py:
```diff
@@ -30,12 +31,25 @@ class CAR:
   {
     2: 8, 64: 8, 65: 8, 72: 8, 73: 8, 280: 8, 281: 8, 282: 8, 290: 8, 312: 8, 313: 8, 314: 8, 315: 8, 316: 8, 326: 8, 372: 8, 544: 8, 545: 8, 546: 8, 554: 8, 557: 8, 576: 8, 577: 8, 722: 8, 801: 8, 802: 8, 803: 8, 805: 8, 808: 8, 811: 8, 826: 8, 837: 8, 838: 8, 839: 8, 842: 8, 912: 8, 915: 8, 940: 8, 1614: 8, 1617: 8, 1632: 8, 1650: 8, 1651: 8, 1657: 8, 1658: 8, 1677: 8, 1722: 8, 1759: 8, 1787: 5 ...
   }],
+  CAR.LEGACY_PREGLOBAL: [{
```
merge all the fingerprints

martinl replied:
do I underestand correctly that I should add all LEGACY_PREGLOBAL unique message counts to one FPv1?

adeebshihadeh replied:
yes

martinl replied:
I tested to combining all Legacy fingerprints to one on subaru-PR-test, it would overlap with existing Outback fingerprint (and prevent upstreaming Outback):

```
./selfdrive/test/test_fingerprints.py
Those two fingerprints are inconsistent SUBARU OUTBACK 2015 - 2017 SUBARU LEGACY 2015 - 2018

2: 8, 208: 8, 209: 4, 210: 8, 211: 7, 212: 8, 316: 8, 320: 8, 321: 8, 324: 8, 328: 8, 329: 8, 336: 2, 338: 8, 342: 8, 352: 8, 353: 8, 354: 8, 356: 8, 358: 8, 359: 8, 392: 8, 554: 8, 604: 8, 640: 8, 642: 8, 644: 8, 805: 8, 864: 8, 865: 8, 866: 8, 872: 8, 880: 8, 881: 8, 882: 8, 884: 8, 885: 8, 977: 8, 1614: 8, 1632: 8, 1657: 8, 1658: 8, 1672: 8, 1722: 8, 1736: 8, 1743: 8, 1745: 8, 1785: 5, 1786: 5, 1787: 5, 1788: 8

2: 8, 208: 8, 209: 4, 210: 8, 211: 7, 212: 8, 316: 8, 320: 8, 321: 8, 324: 8, 328: 8, 329: 8, 336: 2, 338: 8, 342: 8, 352: 8, 353: 8, 354: 8, 356: 8, 358: 8, 359: 8, 392: 8, 554: 8, 604: 8, 640: 8, 642: 8, 644: 8, 805: 8, 864: 8, 865: 8, 866: 8, 872: 8, 880: 8, 881: 8, 882: 8, 884: 8, 885: 8, 977: 8, 1614: 8, 1632: 8, 1640: 8, 1657: 8, 1658: 8, 1672: 8, 1722: 8, 1736: 8, 1743: 8, 1745: 8, 1785: 5, 1786: 5, 1787: 5, 1788: 8, 2015: 8, 2016: 8, 2024: 8

Found 139 individual fingerprints
TEST FAILED
```

martinl replied:
I hand picked the non-overlapping fingerprints from Bugsy's Subaru-PR branch so we would have at least a few FPv1 for each preglobal model, so I would prefer to leave them as is. Afaik there have been no FPv1 false positives on subaru-PR-test

martinl commented:
Rebased to use new signals in Global dbc and also changed driver torque limit to match Global (60). One failed test is due to missing route on the server

martinl commented:
Matching panda safety driver torque change is included in https://github.com/commaai/panda/pull/564

adeebshihadeh commented:
This looks pretty good, but unfortunately it seems like the EPS is different from the global platform, meaning we'll have to rent one to verify the new safety mode. Let's put this behind the dashcamOnly flag to get this merged until we can verify the safety.

martinl commented:
> This looks pretty good, but unfortunately it seems like the EPS is different from the global platform, meaning we'll have to rent one to verify the new safety mode. Let's put this behind the dashcamOnly flag to get this merged until we can verify the safety.

That sounds reasonable, added dashcamOnly to LEGACY_PREGLOBAL

adeebshihadeh commented on selfdrive/car/subaru/subarucan.py:
```diff
@@ -1,19 +1,36 @@
 
-  # counts from 0 to 15 then back to 0 + 16 for enable bit
-  idx = ((frame // steer_step) % 16)
+def create_steering_control(packer, car_fingerprint, apply_steer, frame, steer_step):
```
Let's just split this into different functions for each platform

martinl replied:
done

adeebshihadeh commented on README.md:
```diff
@@ -163,6 +163,7 @@ Community Maintained Cars and Features
 | Subaru    | Crosstrek 2018-19             | EyeSight          | Stock            | 0mph               | 0mph         |
 | Subaru    | Forester 2019                 | EyeSight          | Stock            | 0mph               | 0mph         |
 | Subaru    | Impreza 2017-19               | EyeSight          | Stock            | 0mph               | 0mph         |
+| Subaru    | Legacy 2015-18                | EyeSight          | Stock            | 0mph               | 0mph         |
```
Revert the README change, we'll add it once it's not dashcam only

martinl replied:
done

adeebshihadeh commented on selfdrive/car/subaru/carstate.py:
```diff
@@ -102,64 +115,110 @@ def get_can_parser(CP):
+        ("Signal7", "ES_CruiseThrottle", 0),
+      ]
+
+      checks = [
+        ("ES_DashStatus", 20),
+      ]
```
Add checks for all messages you're reading in carState

martinl replied:
done, I did not include BodyInfo and Dash_State since they are ~0.8Hz and I think 1Hz is minimum

martinl commented:
rebased, tests passed
